source/opt/loop_utils.cpp

```diff
@@ -48,125 +48,249 @@ class LCSSARewriter {
+    // UpdateManagers.
+    void RewriteUse(ir::BasicBlock* bb, ir::Instruction* user,
+                    uint32_t operand_index) {
+      assert(
+          (user->opcode() != SpvOpPhi || bb != GetParent(user)) &&
+          "The root basic block must be the incoming edge if |user| is a phi "
+          "instruction");
+      assert((user->opcode() == SpvOpPhi || bb == GetParent(user)) &&
+             "The root basic block must be the instruction parent if |user| is "
+             "not "
+             "phi instruction");
+
+      ir::Instruction* new_def = GetOrBuildIncoming(bb->id());
+
+      user->SetOperand(operand_index, {new_def->result_id()});
+      rewritten_.insert(user);
+    }
 
-  // Notifies the addition of a phi node built to close the loop.
-  inline void RegisterExitPhi(ir::BasicBlock* bb, ir::Instruction* phi) {
-    bb_to_phi[bb->id()] = phi;
-    rewritten.insert(phi);
-  }
+    // Notifies the addition of a phi node built to close the loop.
+    inline void RegisterExitPhi(ir::BasicBlock* bb, ir::Instruction* phi) {
+      bb_to_phi_[bb->id()] = phi;
+      rewritten_.insert(phi);
+    }
 
-  // In-place update of some managers (avoid full invalidation).
-  inline void UpdateManagers() {
-    opt::analysis::DefUseManager* def_use_mgr = context_->get_def_use_mgr();
-    // Register all new definitions.
-    for (ir::Instruction* insn : rewritten) {
-      def_use_mgr->AnalyzeInstDef(insn);
+    // In-place update of some managers (avoid full invalidation).
+    inline void UpdateManagers() {
+      opt::analysis::DefUseManager* def_use_mgr =
+          base_->context_->get_def_use_mgr();
+      // Register all new definitions.
+      for (ir::Instruction* insn : rewritten_) {
+        def_use_mgr->AnalyzeInstDef(insn);
+      }
+      // Register all new uses.
+      for (ir::Instruction* insn : rewritten_) {
+        def_use_mgr->AnalyzeInstUse(insn);
+      }
     }
-    // Register all new uses.
-    for (ir::Instruction* insn : rewritten) {
-      def_use_mgr->AnalyzeInstUse(insn);
+
+   private:
+    // Return the basic block that |instr| belongs to.
+    ir::BasicBlock* GetParent(ir::Instruction* instr) {
+      return base_->context_->get_instr_block(instr);
     }
-  }
 
- private:
-  // Return the basic block that |instr| belongs to.
-  ir::BasicBlock* GetParent(ir::Instruction* instr) {
-    return context_->get_instr_block(instr);
-  }
+    // Return the new def to use for the basic block |bb_id|.
+    // If |bb_id| does not have a suitable def to use then we:
+    //   - return the common def used by all predecessors;
+    //   - if there is no common def, then we build a new phi instr at the
+    //     beginning of |bb_id| and return this new instruction.
+    ir::Instruction* GetOrBuildIncoming(uint32_t bb_id) {
+      assert(base_->cfg_->block(bb_id) != nullptr && "Unknown basic block");
+
+      ir::Instruction*& incoming_phi = bb_to_phi_[bb_id];
```

Comment:
I think there might be a problem with this uses of bb_to_phi_.

It seems like you will add any phi that uses the id when you "RegisterExitPhi".  It is possible that multiple phis uses the id.  How do you know the correct one, if any of the already existing ones are correct.

An example might be:

```
// bb1
m1=0;
for( int i = 0; i < 10; i++ ) {
   // bb2
   n2 = i;
   if (cond) {
      // bb3
      break;
   }
   // bb4
}
m2 = phi(n2, bb3, 0, bb4);
return n2+m2;
```
In this case, the phi defining `m2` might be added to bb_to_phi_, right?  However, when replacing the uses of `n2` in the return statement, you do not want to use the result of that phi.  You need a new one.  I think there might be an ordering issue with the calls to `RegisterExitPhi`, so it might sometimes get it right, but wrong other times.